File: services/api/src/njz_api/clients/pandascore.py
"""
PandaScore API Client
https://developers.pandascore.co/

Provides Valorant and CS2 data:
- Matches (upcoming, live, completed)
- Teams and rosters
- Players and stats
- Tournaments and series
"""
import os
import httpx
from typing import Any, Optional

PANDASCORE_BASE_URL = "https://api.pandascore.co"


class PandaScoreClient:
    """Async HTTP client for PandaScore API."""

    def __init__(self, api_key: Optional[str] = None):
        self.api_key = api_key or os.environ.get("PANDASCORE_API_KEY", "")
        self.base_url = PANDASCORE_BASE_URL
        self.timeout = 30

    def _headers(self) -> dict:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
        }

    async def _get(self, path: str, params: Optional[dict] = None) -> Any:
        """Make a GET request to PandaScore API."""
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.get(
                f"{self.base_url}{path}",
                headers=self._headers(),
                params=params or {},
            )
            response.raise_for_status()
            return response.json()

    # ── Valorant ─────────────────────────────────────────────

    async def get_valorant_matches(
        self, status: str = "upcoming", page: int = 1, per_page: int = 20
    ) -> list:
        """Get Valorant matches. status: upcoming | running | past"""
        return await self._get(
            "/valorant/matches",
            params={"filter[status]": status, "page": page,
                    "per_page": per_page, "sort": "-begin_at"},
        )

    async def get_valorant_teams(self, page: int = 1, per_page: int = 50) -> list:
        return await self._get(
            "/valorant/teams", params={"page": page, "per_page": per_page}
        )

    async def get_valorant_players(self, page: int = 1, per_page: int = 50) -> list:
        return await self._get(
            "/valorant/players", params={"page": page, "per_page": per_page}
        )

    async def get_valorant_tournaments(self, page: int = 1, per_page: int = 20) -> list:
        return await self._get(
            "/valorant/tournaments/upcoming",
            params={"page": page, "per_page": per_page},
        )

    # ── CS2 ──────────────────────────────────────────────────

    async def get_cs2_matches(
        self, status: str = "upcoming", page: int = 1, per_page: int = 20
    ) -> list:
        """Get CS2 matches. status: upcoming | running | past"""
        return await self._get(
            "/csgo/matches",
            params={"filter[status]": status, "page": page,
                    "per_page": per_page, "sort": "-begin_at"},
        )

    async def get_cs2_teams(self, page: int = 1, per_page: int = 50) -> list:
        return await self._get(
            "/csgo/teams", params={"page": page, "per_page": per_page}
        )

    async def get_cs2_players(self, page: int = 1, per_page: int = 50) -> list:
        return await self._get(
            "/csgo/players", params={"page": page, "per_page": per_page}
        )

    # ── Generic ──────────────────────────────────────────────

    async def get_player_by_id(self, player_id: str, game: str = "valorant") -> dict:
        endpoint = "/valorant/players" if game == "valorant" else "/csgo/players"
        return await self._get(f"{endpoint}/{player_id}")

    async def get_team_by_id(self, team_id: str, game: str = "valorant") -> dict:
        endpoint = "/valorant/teams" if game == "valorant" else "/csgo/teams"
        return await self._get(f"{endpoint}/{team_id}")

    # ── Player Stats ──────────────────────────────────────────

    async def get_valorant_player_stats(
        self, player_id: int, page: int = 1, per_page: int = 50
    ) -> list:
        """Fetch recent per-match stats for a Valorant player."""
        return await self._get(
            f"/valorant/players/{player_id}/stats",
            params={"page": page, "per_page": per_page},
        )

    async def get_cs2_player_stats(
        self, player_id: int, page: int = 1, per_page: int = 50
    ) -> list:
        """Fetch recent per-match stats for a CS2 player."""
        return await self._get(
            f"/csgo/players/{player_id}/stats",
            params={"page": page, "per_page": per_page},
        )


# Module-level singleton — use this in route handlers
pandascore = PandaScoreClient()

File: apps/web/src/services/pandascore.ts
/**
 * PandaScore API Service
 * Real-time esports data for ROTAS Station
 */

import { useState, useEffect } from 'react';

const PANDASCORE_API_URL = 'https://api.pandascore.co';
const API_TOKEN = import.meta.env.VITE_PANDASCORE_TOKEN || '';

// Types
export interface Tournament {
  id: number;
  name: string;
  slug: string;
  begin_at: string;
  end_at: string;
  prizepool: string | null;
  serie_id: number;
  serie: {
    name: string;
    full_name: string;
  };
  league: {
    name: string;
    image_url: string | null;
  };
  tier: string;
}

export interface Match {
  id: number;
  name: string;
  slug: string;
  status: string;
  begin_at: string;
  end_at: string | null;
  number_of_games: number;
  tournament_id: number;
  opponents: Array<{
    opponent: {
      id: number;
      name: string;
      acronym: string;
      image_url: string | null;
    };
  }>;
  results: Array<{
    score: number;
    team_id: number;
  }>;
  winner_id: number | null;
  winner_type: string | null;
}

export interface Team {
  id: number;
  name: string;
  slug: string;
  acronym: string;
  image_url: string | null;
  location: string;
  players: Player[];
}

export interface Player {
  id: number;
  name: string;
  slug: string;
  first_name: string | null;
  last_name: string | null;
  nationality: string;
  image_url: string | null;
  role: string | null;
  age: number | null;
  team_id: number;
}

// API Client
class PandaScoreClient {
  private baseURL: string;
  private token: string;

  constructor(token: string) {
    this.baseURL = PANDASCORE_API_URL;
    this.token = token;
  }

  private async fetch(endpoint: string, params: Record<string, string> = {}): Promise<any> {
    const queryParams = new URLSearchParams(params);
    const url = `${this.baseURL}${endpoint}?${queryParams.toString()}`;
    
    try {
      const response = await fetch(url, {
        headers: {
          'Authorization': `Bearer ${this.token}`,
          'Accept': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error(`API Error: ${response.status} ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('PandaScore API Error:', error);
      throw error;
    }
  }

  // Valorant endpoints
  async getValorantTournaments(): Promise<Tournament[]> {
    if (!this.token) return [];
    return this.fetch('/valorant/tournaments', { 'filter[tier]': 's,a', 'sort': '-begin_at' });
  }

  async getValorantMatches(status: string = 'running'): Promise<Match[]> {
    if (!this.token) return [];
    return this.fetch('/valorant/matches', { 'filter[status]': status, 'sort': '-begin_at' });
  }

  async getValorantTeams(): Promise<Team[]> {
    if (!this.token) return [];
    return this.fetch('/valorant/teams', { 'sort': 'name' });
  }

  async getValorantPlayers(): Promise<Player[]> {
    if (!this.token) return [];
    return this.fetch('/valorant/players', { 'sort': 'name' });
  }
}

// Create client instance
export const pandaScoreClient = new PandaScoreClient(API_TOKEN);

// React Hooks for data fetching
export function useValorantTournaments() {
  const [tournaments, setTournaments] = useState<Tournament[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    let mounted = true;

    const fetchTournaments = async () => {
      try {
        setLoading(true);
        // If no token, use mock data for demo
        if (!API_TOKEN) {
          if (mounted) {
            setTournaments(getMockTournaments());
          }
          return;
        }
        const data = await pandaScoreClient.getValorantTournaments();
        if (mounted) {
          setTournaments(data);
        }
      } catch (err) {
        if (mounted) {
          setError(err as Error);
          // Fallback to mock data on error
          setTournaments(getMockTournaments());
        }
      } finally {
        if (mounted) {
          setLoading(false);
        }
      }
    };

    fetchTournaments();
    return () => { mounted = false; };
  }, []);

  return { tournaments, loading, error };
}

export function useValorantMatches(status: string = 'running') {
  const [matches, setMatches] = useState<Match[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    let mounted = true;

    const fetchMatches = async () => {
      try {
        setLoading(true);
        if (!API_TOKEN) {
          if (mounted) {
            setMatches(getMockMatches());
          }
          return;
        }
        const data = await pandaScoreClient.getValorantMatches(status);
        if (mounted) {
          setMatches(data);
        }
      } catch (err) {
        if (mounted) {
          setError(err as Error);
          setMatches(getMockMatches());
        }
      } finally {
        if (mounted) {
          setLoading(false);
        }
      }
    };

    fetchMatches();
    return () => { mounted = false; };
  }, [status]);

  return { matches, loading, error };
}

export function useValorantTeams() {
  const [teams, setTeams] = useState<Team[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    let mounted = true;

    const fetchTeams = async () => {
      try {
        setLoading(true);
        if (!API_TOKEN) {
          if (mounted) {
            setTeams(getMockTeams());
          }
          return;
        }
        const data = await pandaScoreClient.getValorantTeams();
        if (mounted) {
          setTeams(data);
        }
      } catch (err) {
        if (mounted) {
          setError(err as Error);
          setTeams(getMockTeams());
        }
      } finally {
        if (mounted) {
          setLoading(false);
        }
      }
    };

    fetchTeams();
    return () => { mounted = false; };
  }, []);

  return { teams, loading, error };
}

// Mock data for demo/fallback
function getMockTournaments(): Tournament[] {
  return [
    {
      id: 1,
      name: 'VCT 2024 Champions Tour',
      slug: 'vct-2024-champions-tour',
      begin_at: '2024-01-15T00:00:00Z',
      end_at: '2024-08-25T00:00:00Z',
      prizepool: '$2,250,000',
      serie_id: 1,
      serie: { name: 'VCT Americas', full_name: 'VCT 2024 Americas' },
      league: { name: 'Riot Games', image_url: null },
      tier: 's',
    },
    {
      id: 2,
      name: 'VCT 2024 Masters Madrid',
      slug: 'vct-2024-masters-madrid',
      begin_at: '2024-03-14T00:00:00Z',
      end_at: '2024-03-24T00:00:00Z',
      prizepool: '$500,000',
      serie_id: 2,
      serie: { name: 'VCT Masters', full_name: 'VCT 2024 Masters' },
      league: { name: 'Riot Games', image_url: null },
      tier: 's',
    },
    {
      id: 3,
      name: 'VCT Challengers NA',
      slug: 'vct-challengers-na',
      begin_at: '2024-02-01T00:00:00Z',
      end_at: '2024-05-15T00:00:00Z',
      prizepool: '$100,000',
      serie_id: 3,
      serie: { name: 'VCL', full_name: 'VCT Challengers' },
      league: { name: 'Riot Games', image_url: null },
      tier: 'a',
    },
  ];
}

function getMockMatches(): Match[] {
  return [
    {
      id: 1,
      name: 'Sentinels vs Cloud9',
      slug: 'sentinels-vs-cloud9',
      status: 'finished',
      begin_at: '2024-03-20T20:00:00Z',
      end_at: '2024-03-20T22:30:00Z',
      number_of_games: 3,
      tournament_id: 1,
      opponents: [
        { opponent: { id: 1, name: 'Sentinels', acronym: 'SEN', image_url: null } },
        { opponent: { id: 2, name: 'Cloud9', acronym: 'C9', image_url: null } },
      ],
      results: [
        { score: 2, team_id: 1 },
        { score: 1, team_id: 2 },
      ],
      winner_id: 1,
      winner_type: 'Team',
    },
    {
      id: 2,
      name: 'NRG vs LOUD',
      slug: 'nrg-vs-loud',
      status: 'upcoming',
      begin_at: '2024-04-05T22:00:00Z',
      end_at: null,
      number_of_games: 3,
      tournament_id: 1,
      opponents: [
        { opponent: { id: 3, name: 'NRG', acronym: 'NRG', image_url: null } },
        { opponent: { id: 4, name: 'LOUD', acronym: 'LOUD', image_url: null } },
      ],
      results: [],
      winner_id: null,
      winner_type: null,
    },
  ];
}

function getMockTeams(): Team[] {
  return [
    {
      id: 1,
      name: 'Sentinels',
      slug: 'sentinels',
      acronym: 'SEN',
      image_url: null,
      location: 'US',
      players: [
        { id: 1, name: 'TenZ', slug: 'tenz', first_name: 'Tyson', last_name: 'Ngo', nationality: 'CA', image_url: null, role: 'duelist', age: 23, team_id: 1 },
        { id: 2, name: 'zekken', slug: 'zekken', first_name: 'Zachary', last_name: 'Patrone', nationality: 'US', image_url: null, role: 'duelist', age: 21, team_id: 1 },
      ],
    },
    {
      id: 2,
      name: 'Cloud9',
      slug: 'cloud9',
      acronym: 'C9',
      image_url: null,
      location: 'US',
      players: [
        { id: 3, name: 'vanity', slug: 'vanity', first_name: 'Anthony', last_name: 'Malaspina', nationality: 'US', image_url: null, role: 'controller', age: 24, team_id: 2 },
      ],
    },
  ];
}

export default pandaScoreClient;
